ls1intum/Artemis
# Notification popup

### Checklist

### Motivation and Context
If students are not on the course overview, they will not see when a new quiz exercise starts. We want to notify students about a started quiz exercise everywhere.

### Description
We use a popup in the right bottom corner to inform students about a started quiz exercise. The popup is shown on every page. Therefore I introduce a couple of changes:
- To prevent unnecessary server load we derive the corresponding notification from the already existing quiz updates that are received via websocket. We do not send the quiz started notification via websocket.
- At the moment the subscription to the websocket messages for notifications takes mainly place in the `course-management.service`. It makes more sense to place these subscriptions in the components that actually consume the notifications (`notification-sidebar.component` and `notification-popup.component`).
- The `course-management.service` now only provides a list of `coursesForNotifications`. The `notification.service` uses this list to subscribe to notification updates and quiz updates (if not already subscribed).
- The `coursesForNotifications` will be filled with courses received from existing requests from the `course-management.service`. I've added a REST route that retrieves the courses for notifications with minimal payload.
- The new `GET /courses/for-notifications` request is only triggered if no courses have been retrieved up by the course management service yet.
- The notification will only be displayed if the user is not already on the target page (or the live participation page)

### Steps for Testing
1. Log in to Artemis with *artemis_test_user1*
2. Navigate wherever you want
3. Execute steps 1-5 in *browser/tab 2*
4. The notification should now popup in this tab
5. Click on the notification
6. You should be redirected to the quiz's participation page

**Browser/Tab 2**
1. Log in to Artemis with your account
2. Navigate to Course Administration
3. Create a quiz
4. Set the quiz visible
5. Start the quiz

### Test Coverage
**Client**
- `notification-sidebar.component.ts` 88.89% -> **88.24%**
- `notification-popup.component.ts` **97.3%**
- `course-management.service.ts` 31.88% -> **32.16%**
- `notification.service.ts` 32.88% -> **47.37%**

**Server**
Flaky test prevents generation of coverage report.

### Screenshots
![notification-popup](https://user-images.githubusercontent.com/15168372/83081603-5ecdea80-a081-11ea-9942-e804007b658c.gif)
